src/oop/static_modifier/medium/q4/MainTest.java

```diff
@@ -0,0 +1,130 @@
+        assertTrue(testClass.hasField(firstDoubleName, double.class, new String[]{"private", "static"}),
+                String.format(modificationErrorMessage,firstDoubleName));
+        assertTrue(testClass.hasField(secondDoubleName, double.class, new String[]{"private", "static"}),
+                String.format(modificationErrorMessage, secondDoubleName));
+        assertTrue(testClass.hasField(firstIntName, int.class, new String[]{"private", "final", "static"}),
+                String.format(modificationErrorMessage, firstIntName));
+        assertTrue(testClass.hasField(secondIntName, int.class, new String[]{"private", "final", "static"}),
+                String.format(modificationErrorMessage, secondIntName));
+        assertTrue(testClass.hasField(arrayName, double[].class, new String[]{"private"}),
+                String.format(modificationErrorMessage, arrayName));
+        modificationErrorMessage = "You have modified the provided portions of " + className + " class for the %s constructor. Please revert them to the original state.";
+        for (int i = 0; i < 10; i++) {
+            double[] test = generateRandomArray(0.5, 20.0, ((int) (Math.random() * 20)));
+            Object[][] arguments = {{test, double[].class}};
+            Object classInstance = testClass.createInstance(arguments);
+            _assertArrayEquals(test, testClass.getFieldValue(classInstance, arrayName),
+                    String.format(modificationErrorMessage, className));
+            _assertEquals(initialValue1, testClass.getFieldValue(null, firstDoubleName),
+                    String.format(modificationErrorMessage, className));
+            _assertEquals(initialValue2, testClass.getFieldValue(null, secondDoubleName),
+                    String.format(modificationErrorMessage, className));
+            _assertEquals(typeValue1, testClass.getFieldValue(null, firstIntName),
+                    String.format(modificationErrorMessage, className));
+            _assertEquals(typeValue2, testClass.getFieldValue(null, secondIntName),
+                    String.format(modificationErrorMessage, className));
+
+        }
+    }
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of(new double[]{initialValue1, initialValue2, initialValue1, initialValue2, initialValue1}),
+                Arguments.of(new double[]{initialValue1 + 1, initialValue2 + 1, initialValue1 + 1, initialValue2 + 1, initialValue1 + 1}),
+                Arguments.of(new double[]{initialValue1 - 1, initialValue2 - 1, initialValue1 - 1, initialValue2 - 1, initialValue1 - 1}),
+                Arguments.of(generateRandomArray(0.5, 20.0, 10)),
+                Arguments.of(generateRandomArray(0.5, 20.0, 10)),
+                Arguments.of(generateRandomArray(0.5, 20.0, 10))
```

Comment:
You should redo these later cases. Since you are using doubles, you should not do initial value + 1, it should be + 0.01 or whatever your `delta` value is in the assert (see my comment below).

Reply:
You should also add a test where there is enough of liquid 1, but not liquid 2 and vice versa.

Reply:
I don't think having the random values here adds too much.